### PR TITLE
Added possibility to add/delete tags to category & discount

### PR DESCRIPTION
**Pull request by VolodymyrBushko**

### CODE REVIEW

File: src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java [hunk 1]
```diff
@@ -39,6 +41,18 @@ public CategoryResponseVO update(@PathVariable Long id, @Valid @RequestBody Cate
         return categoryService.update(id, request);
     }
 
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PostMapping("/{id}/tags")
+    public List<TagResponseVO> addTags(@PathVariable Long id, @Valid @RequestBody List<TagRequestVO> tagRequest) {
+        return categoryService.addTags(id, tagRequest);
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java]: rename to categoryId?

File: src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java [hunk 2]
```diff
@@ -39,6 +41,18 @@ public CategoryResponseVO update(@PathVariable Long id, @Valid @RequestBody Cate
...
+    }
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @DeleteMapping("/{id}/tags")
+    public void deleteTags(@PathVariable Long id, @Valid @RequestBody List<Long> tagIds) {
+        categoryService.deleteTags(id, tagIds);
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java]: validate whether this tag is already used in discount

File: src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java [hunk 3]
```diff
@@ -55,6 +62,44 @@ public CategoryResponseVO update(Long id, CategoryRequestVO request) {
...
+        newTags.forEach(e -> {
+            if (categoryTags.contains(e)) {
+                throw new EntityAlreadyExistsException(Tag.class, "name", e.getName());
+            }
+        });
+
+        categoryTags.addAll(newTags);
+        categoryRepository.save(category);
+
+        return category.getTags()
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java]: let's return category response with the list of tags

File: src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java [hunk 4]
```diff
@@ -55,6 +62,44 @@ public CategoryResponseVO update(Long id, CategoryRequestVO request) {
...
+
+        List<Tag> newTags = tagRequest
+                .stream()
+                .peek(e -> e.setCategoryId(category.getId()))
+                .map(tagMapper::toEntity)
+                .collect(Collectors.toList());
+
+        List<Tag> categoryTags = category.getTags();
+
+        newTags.forEach(e -> {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java]: let's move to validator

File: src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java [hunk 5]
```diff
@@ -43,6 +45,18 @@ public DiscountResponseVO update(@PathVariable Long id, @Valid @RequestBody Disc
         return discountService.update(id, request);
     }
 
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @PostMapping("/{id}/tags")
+    public List<TagResponseVO> addTags(@PathVariable Long id, @Valid @RequestBody List<Long> tagIds) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java]: let's link discount with tags during the discount creation (list of tag ids should be in the discount request)

File: src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java [hunk 6]
```diff
@@ -43,6 +45,18 @@ public DiscountResponseVO update(@PathVariable Long id, @Valid @RequestBody Disc
...
+        return discountService.addTags(id, tagIds);
+    }
+
+    @PreAuthorize("hasAuthority('ADMIN')")
+    @DeleteMapping("/{id}/tags")
+    public void deleteTags(@PathVariable Long id, @Valid @RequestBody List<Long> tagIds) {
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java]: not needed

File: src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java (hunk 1 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java]: done

File: src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java (hunk 3 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java]: done

File: src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java (hunk 4 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/category/service/CategoryService.java]: done

File: src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java (hunk 5 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java]: done

File: src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java (hunk 6 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/discount/controller/DiscountController.java]: removed

File: src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java (hunk 2 above)
Comment (VolodymyrBushko) [reply src/main/java/com/exadel/discountwebapp/category/controller/CategoryController.java]: done

File: src/main/java/com/exadel/discountwebapp/category/validator/CategoryValidator.java [hunk 7]
```diff
@@ -17,10 +23,35 @@ public void validate(CategoryRequestVO request) {
...
+    public void checkDuplicateTag(Category category, List<Tag> newTags) {
+        newTags.forEach(e -> {
+            if (category.getTags().contains(e)) {
+                throw new EntityAlreadyExistsException(Tag.class, "name", e.getName());
+            }
+        });
+    }
+
+    public void checkTagAlreadyUsedInDiscount(Category category, List<Long> tagIds) {
+        List<Discount> discounts = category.getDiscounts();
```
Comment (ahlukhouski2) [inline src/main/java/com/exadel/discountwebapp/category/validator/CategoryValidator.java:42]: Here we are retrieving all discounts of the category.
The amount of data retrieved from the database can be too large.
Probably this check can be performed via a single SQL query without retrieving all the data from the database.
But let's make this optimization later.

**Review by ahlukhouski2 — APPROVED** [review]

**Pull request merged**